File: shared/agent/src/protocol/agent.protocol.notifications.ts
"use strict";
import { NotificationType, TextDocumentIdentifier } from "vscode-languageserver-protocol";
import { Document, GetMyPullRequestsResponse } from "./agent.protocol";
import { LoginSuccessResponse, TokenLoginRequest } from "./agent.protocol.auth";
import { CodemarkPlus } from "./agent.protocol.codemarks";
import { ThirdPartyProviders } from "./agent.protocol.providers";
import {
	CSApiCapabilities,
	CSCompany,
	CSLastReads,
	CSMarker,
	CSMarkerLocations,
	CSMePreferences,
	CSPost,
	CSRepository,
	CSReview,
	CSStream,
	CSTeam,
	CSUser
} from "./api.protocol";

export interface RestartRequiredNotification {}

export const RestartRequiredNotificationType = new NotificationType<
	RestartRequiredNotification,
	void
>("codestream/restartRequired");

export enum ConnectionStatus {
	Disconnected = "disconnected",
	Reconnected = "reconnected",
	Reconnecting = "reconnecting"
}

export interface DidChangeConnectionStatusNotification {
	reset?: boolean;
	status: ConnectionStatus;
}

export const DidChangeConnectionStatusNotificationType = new NotificationType<
	DidChangeConnectionStatusNotification,
	void
>("codestream/didChangeConnectionStatus");

export enum ChangeDataType {
	Codemarks = "codemarks",
	Commits = "commits",
	Companies = "companies",
	Documents = "documents",
	MarkerLocations = "markerLocations",
	Markers = "markers",
	Posts = "posts",
	Preferences = "preferences",
	PullRequests = "pullRequests",
	Repositories = "repos",
	Reviews = "reviews",
	Streams = "streams",
	Teams = "teams",
	Unreads = "unreads",
	Users = "users",
	Providers = "providers",
	ApiCapabilities = "apiCapabilities",
	Workspace = "workspace"
}

export interface CodemarksChangedNotification {
	type: ChangeDataType.Codemarks;
	data: CodemarkPlus[];
}

export interface CompaniesChangedNotification {
	type: ChangeDataType.Companies;
	data: CSCompany[];
}

export interface MarkerLocationsChangedNotification {
	type: ChangeDataType.MarkerLocations;
	data: CSMarkerLocations[];
}

export interface MarkersChangedNotification {
	type: ChangeDataType.Markers;
	data: CSMarker[];
}

export interface PostsChangedNotification {
	type: ChangeDataType.Posts;
	data: CSPost[];
}

export interface PreferencesChangedNotification {
	type: ChangeDataType.Preferences;
	data: CSMePreferences;
}

export interface PullRequestsChangedData {
	queryName: string;
	pullRequest: GetMyPullRequestsResponse;
}

export interface PullRequestsChangedNotification {
	type: ChangeDataType.PullRequests;
	data: PullRequestsChangedData[];
}

export interface RepositoriesChangedNotification {
	type: ChangeDataType.Repositories;
	data: CSRepository[];
}

export interface ReviewsChangedNotification {
	type: ChangeDataType.Reviews;
	data: CSReview[];
}

export interface StreamsChangedNotification {
	type: ChangeDataType.Streams;
	data: CSStream[];
}

export interface TeamsChangedNotification {
	type: ChangeDataType.Teams;
	data: CSTeam[];
}

export interface Unreads {
	lastReads: CSLastReads;
	mentions: { [streamId: string]: number };
	unreads: { [streamId: string]: number };
	totalMentions: number;
	totalUnreads: number;
}

export interface UnreadsChangedNotification {
	type: ChangeDataType.Unreads;
	data: Unreads;
}

export interface UsersChangedNotification {
	type: ChangeDataType.Users;
	data: CSUser[];
}

export interface ProvidersChangedNotification {
	type: ChangeDataType.Providers;
	data: ThirdPartyProviders;
}

export interface ApiCapabilitiesChangedNotification {
	type: ChangeDataType.ApiCapabilities;
	data: CSApiCapabilities;
}

export interface DocumentData {
	reason: "saved" | "changed" | "removed";
	document: Document;
}

export interface DocumentsChangedNotification {
	type: ChangeDataType.Documents;
	data: DocumentData;
}

export interface CommitsChangedData {
	type: string;
	path: string;
	repo: {
		id: string | undefined;
		path: string;
		normalizedPath: string;
	};
}

export interface WorkspaceChangedData {}

export interface CommitsChangedNotification {
	type: ChangeDataType.Commits;
	data: CommitsChangedData;
}

export interface WorkspaceChangedNotification {
	type: ChangeDataType.Workspace;
	data: WorkspaceChangedData;
}

export type DidChangeDataNotification =
	| CodemarksChangedNotification
	| CompaniesChangedNotification
	| MarkerLocationsChangedNotification
	| MarkersChangedNotification
	| PostsChangedNotification
	| PreferencesChangedNotification
	| PullRequestsChangedNotification
	| RepositoriesChangedNotification
	| ReviewsChangedNotification
	| StreamsChangedNotification
	| TeamsChangedNotification
	| UnreadsChangedNotification
	| UsersChangedNotification
	| ProvidersChangedNotification
	| ApiCapabilitiesChangedNotification
	| DocumentsChangedNotification
	| CommitsChangedNotification
	| WorkspaceChangedNotification;

export const DidChangeDataNotificationType = new NotificationType<DidChangeDataNotification, void>(
	"codestream/didChangeData"
);

export interface DidChangeDocumentMarkersNotification {
	textDocument: TextDocumentIdentifier;
	reason: "document" | "codemarks" | "pullRequestComments";
}

export const DidChangeDocumentMarkersNotificationType = new NotificationType<
	DidChangeDocumentMarkersNotification,
	void
>("codestream/didChangeDocumentMarkers");

export interface DidChangePullRequestCommentsNotification {
	pullRequestId: string;
	commentId?: string;
	filePath?: string;
}

export const DidChangePullRequestCommentsNotificationType = new NotificationType<
	DidChangePullRequestCommentsNotification,
	void
>("codestream/didChangePullRequestComments");

export enum VersionCompatibility {
	Compatible = "ok",
	CompatibleUpgradeAvailable = "outdated",
	CompatibleUpgradeRecommended = "deprecated",
	UnsupportedUpgradeRequired = "incompatible",
	Unknown = "unknownVersion"
}

export interface DidChangeVersionCompatibilityNotification {
	compatibility: VersionCompatibility;
	downloadUrl: string;
	version: string | undefined;
}

export const DidChangeVersionCompatibilityNotificationType = new NotificationType<
	DidChangeVersionCompatibilityNotification,
	void
>("codestream/didChangeVersionCompatibility");

export enum ApiVersionCompatibility {
	ApiCompatible = "apiCompatible",
	ApiUpgradeRecommended = "apiUpgradeRecommended",
	ApiUpgradeRequired = "apiUpgradeRequired"
}

export interface DidChangeApiVersionCompatibilityNotification {
	compatibility: ApiVersionCompatibility;
	version: string;
	missingCapabilities?: CSApiCapabilities;
}

export const DidChangeApiVersionCompatibilityNotificationType = new NotificationType<
	DidChangeApiVersionCompatibilityNotification,
	void
>("codestream/didChangeApiVersionCompatibility");

export enum LogoutReason {
	Token = "token",
	Unknown = "unknown",
	UnsupportedVersion = "unsupportedVersion",
	UnsupportedApiVersion = "unsupportedApiVersion"
}

export interface DidLogoutNotification {
	reason: LogoutReason;
}

export const DidLogoutNotificationType = new NotificationType<DidLogoutNotification, void>(
	"codestream/didLogout"
);

export interface DidLoginNotification {
	data: LoginSuccessResponse;
}
export const DidLoginNotificationType = new NotificationType<DidLoginNotification, void>(
	"codestream/didLogin"
);

export const DidStartLoginNotificationType = new NotificationType<void, void>(
	"codestream/didStartLogin"
);

export const DidFailLoginNotificationType = new NotificationType<void, void>(
	"codestream/didFailLogin"
);

export type DidEncounterMaintenanceModeNotification = TokenLoginRequest;

export const DidEncounterMaintenanceModeNotificationType = new NotificationType<
	DidEncounterMaintenanceModeNotification,
	void
>("codestream/didEncounterMaintenanceMode");

export interface DidChangeServerUrlNotification {
	serverUrl: string;
}

export const DidChangeServerUrlNotificationType = new NotificationType<
	DidChangeServerUrlNotification,
	void
>("codestream/didChangeServerUrl");

export const AgentInitializedNotificationType = new NotificationType<void, void>(
	"codestream/agentInitialized"
);

export interface UserDidCommitNotification {
	sha: string;
}

export const UserDidCommitNotificationType = new NotificationType<UserDidCommitNotification, void>(
	"codestream/userDidCommit"
);

File: vs/src/CodeStream.VisualStudio.Core/Models/Agent.Protocol.Notifications.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Collections.Generic;

namespace CodeStream.VisualStudio.Core.Models {
	public class DidChangeConnectionStatusNotification {
		public bool? Reset { get; set; }
		public ConnectionStatus Status { get; set; }
	}

	public class DidChangeConnectionStatusNotificationType : NotificationType<DidChangeConnectionStatusNotification> {
		public DidChangeConnectionStatusNotificationType(DidChangeConnectionStatusNotification @params) {
			Params = @params;
		}

		public const string MethodName = "codestream/didChangeConnectionStatus";
		public override string Method => MethodName;
	}

	public class DidChangeUserPreferencesData {
		public bool? CodemarksShowPRComments { get; set; }
		public bool? CodemarksHideReviews { get; set; }
		public bool? CodemarksHideResolved { get; set; }
		public bool? CodemarksShowArchived { get; set; }
	}

	public class DidChangeUserPreferencesEvent {
		public string Type { get; set; } = "preferences";
		public DidChangeUserPreferencesData Data { get; set; }
	} 
	
	public class DidChangeDataNotificationTypeParams {
		 
	}
	public class DidChangeDataNotificationType : NotificationType<DidChangeDataNotificationTypeParams> {
		private readonly JToken _token;

		public DidChangeDataNotificationType(JToken token) {
			_token = token;
		}		 

		public const string MethodName = "codestream/didChangeData";
		public override string Method => MethodName;

		public override string AsJson() {
			return CustomNotificationPayload.Create(Method, _token);
		}
	}

	[JsonConverter(typeof(CamelCaseStringEnumConverter))]
	public enum ApiVersionCompatibility {
		ApiCompatible,
		ApiUpgradeRecommended,
		ApiUpgradeRequired
	}

	// not actually used -- JToken is used so we dont have to deserialze, then serialize
	public class DidChangeApiVersionCompatibilityNotification {
		public ApiVersionCompatibility Compatibility { get; set; }
		public string Version { get; set; }
		public CSApiCapabilities MissingCapabilities { get; set; }
	}

	public class DidChangeApiVersionCompatibilityNotificationType : NotificationType<DidChangeApiVersionCompatibilityNotification> {
		private readonly JToken _token;
		public DidChangeApiVersionCompatibilityNotificationType(JToken token) {
			_token = token;
		}
		public const string MethodName = "codestream/didChangeApiVersionCompatibility";
		public override string Method => MethodName;
		public override string AsJson() {
			return CustomNotificationPayload.Create(Method, _token);
		}
	}

	public class CSApiCapabilities : Dictionary<string, CSApiCapability> {

	}

	public class CSApiCapability {
		public string Description { get; set; }
		public string Url { get; set; }
		public string Version { get; set; }
	}


	[JsonConverter(typeof(CamelCaseStringEnumConverter))]
	public enum ChangeReason {
		Document,
		Codemarks
	}

	public class DidChangeDocumentMarkersNotification {
		public TextDocumentIdentifier TextDocument { get; set; }
		public ChangeReason? Reason { get; set; }
	}

	public class DidChangeDocumentMarkersNotificationType : NotificationType<DidChangeDocumentMarkersNotification> {
		public const string MethodName = "codestream/didChangeDocumentMarkers";
		public override string Method => MethodName;
	}

	//export enum VersionCompatibility
	//{
	//    Compatible = "ok",
	//    CompatibleUpgradeAvailable = "outdated",
	//    CompatibleUpgradeRecommended = "deprecated",
	//    UnsupportedUpgradeRequired = "incompatible",
	//    Unknown = "unknownVersion"
	//}

	public class DidChangeVersionCompatibilityNotification {
		public string Compatibility { get; set; }
		public string DownloadUrl { get; set; }
		public string Version { get; set; }
	}

	public class DidChangeVersionCompatibilityNotificationType : NotificationType<DidChangeVersionCompatibilityNotification> {
		private readonly JToken _token;

		public DidChangeVersionCompatibilityNotificationType(JToken token) {
			_token = token;
		}

		public const string MethodName = "codestream/didChangeVersionCompatibility";
		public override string Method => MethodName;

		public override string AsJson() {
			return CustomNotificationPayload.Create(Method, _token);
		}
	}

	public class DidLogoutNotification {
		public LogoutReason Reason { get; set; }
	}

	public class DidLogoutNotificationType : NotificationType<DidLogoutNotification> {
		public const string MethodName = "codestream/didLogout";
		public override string Method => MethodName;
	}


	public class DidLoginNotification {
		public LoginSuccessResponse Data { get; set; }
	}

	public class DidLoginNotificationType : NotificationType<DidLoginNotification> {
		public const string MethodName = "codestream/didLogin";
		public override string Method => MethodName;
	}

	public class DidStartLoginNotification { }

	public class DidStartLoginNotificationType : NotificationType<DidStartLoginNotification> {

		public const string MethodName = "codestream/didStartLogin";
		public override string Method => MethodName;
	}

	public class DidFailLoginNotification { }
	public class DidFailLoginNotificationType : NotificationType<DidFailLoginNotification> {
		public const string MethodName = "codestream/didFailLogin";
		public override string Method => MethodName;
	}

	public class OtcLoginRequest {
		public string Code { get; set; }
		public string TeamId { get; set; }
		public string Team { get; set; }
		public bool? Alias { get; set; }
	}
	
	public class RestartRequiredNotification { }

	public class RestartRequiredNotificationType : NotificationType<RestartRequiredNotification> {
		public const string MethodName = "codestream/restartRequired";
		public override string Method => MethodName;
	}

	public class DidEncounterMaintenanceModeNotification { }

	public class DidEncounterMaintenanceModeNotificationType : NotificationType<DidEncounterMaintenanceModeNotification> {
		public const string MethodName = "codestream/didEncounterMaintenanceMode";
		public override string Method => MethodName;

		private readonly JToken _token;

		public DidEncounterMaintenanceModeNotificationType(JToken token) {
			_token = token;
		}

		public override string AsJson() {
			return CustomNotificationPayload.Create(Method, _token);
		}
	}

	public class DidChangeServerUrlNotification {
		public string ServerUrl { get; set; }
	}

	public class DidChangeServerUrlNotificationType : NotificationType<DidChangeServerUrlNotification> {
		public const string MethodName = "codestream/didChangeServerUrl";
		public override string Method => MethodName;

		private readonly JToken _token;

		public DidChangeServerUrlNotificationType(JToken token) {
			_token = token;
		}

		public override string AsJson() {
			return CustomNotificationPayload.Create(Method, _token);
		}
	}
}
